src/ruby/puppetserver-lib/puppet/server/execution.rb

```diff
@@ -6,32 +6,43 @@
 class Puppet::Server::Execution
   def self.initialize_execution_stub
     Puppet::Util::ExecutionStub.set do |command, options, stdin, stdout, stderr|
-      if command.is_a?(Array)
-        binary = command.first
-        args = command[1..-1]
-      else
-        binary = command
-        args = nil
-      end
-
-      # TODO - options is currently ignored - https://tickets.puppetlabs.com/browse/SERVER-74
+      # TODO - options relating to stdout/stderr are not yet handled, see SERVER-74 and
+      #  PUP-6640
 
       # We're going to handle STDIN/STDOUT/STDERR in java, so we don't need
       # them here.  However, Puppet::Util::Execution.execute doesn't close them
       # for us, so we have to do that now.
       [stdin, stdout, stderr].each { |io| io.close rescue nil }
 
-      execute(binary, args)
+      execute(command, options)
     end
   end
 
-  def self.execute(command, args = nil)
+  private
+  def self.execute(command, options)
+    if command.is_a?(Array)
+      orig_command_str = command.join(" ")
+      binary = command.first
+      args = command[1..-1]
+    else
+      orig_command_str = command
+      binary = command
+      args = nil
+    end
+
+
     if args && !args.empty?
-      result = ShellUtils.executeCommand(command, args.to_java(:string))
+      result = ShellUtils.executeCommand(binary, args.to_java(:string))
     else
-      result = ShellUtils.executeCommand(command)
+      result = ShellUtils.executeCommand(binary)
+    end
+
+    # TODO - not all options from Puppet::Util::Execution are supported yet, see SERVER-74
+
+    if options[:failonfail] and result.exit_code != 0
+      raise Puppet::ExecutionFailure, "Execution of '#{orig_command_str}' returned #{result.exit_code}: #{result.output.strip}"
     end
 
-    Puppet::Util::Execution::ProcessOutput.new(result.getOutput, result.getExitCode)
+    Puppet::Util::Execution::ProcessOutput.new(result.output, result.exit_code)
```

Comment:
Welcome to the magic of JRuby.  For POJO-style java objects, if you have `getCamelCaseProperty` methods, JRuby supports accessing them via `camel_case_property`.  I thought it would be better to go with the ruby idioms in ruby code, but if people think it's more confusing this way I don't mind changing them back.

FWIW I believe these changes passed travis with no trouble yesterday (and the specs pass with these changes locally).